accepted/2024/end-of-life-packages.md

```diff
@@ -0,0 +1,110 @@
+# End-of-life (EOL) packages and package lifecycle
+<!-- Replace `Title` with an appropriate title for your design -->
+
+- [Jon Douglas](https://github.com/jondouglas) <!-- GitHub username link -->
+- [#9638](https://github.com/NuGet/NuGetGallery/issues/9638), [#9837](https://github.com/NuGet/NuGetGallery/issues/9837) <!-- GitHub Issue link -->
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+Today, there exists two major package status mechanisms for understanding if a package is [vulnerable](https://learn.microsoft.com/en-us/nuget/concepts/auditing-packages) and/or [deprecated](https://learn.microsoft.com/en-us/nuget/nuget-org/deprecate-packages).
+
+Recently, the NuGet team has shipped functionality allowing package authors to include markdown READMEs to contain a hodgepodge of information relating to the package such as installation instructions, getting started examples, contributing tips, and less often seen is how the package is supported through an official or unofficial support policy.
+
+To make these package statuses clearer to package consumers and authors, this proposal introduces a new concept to indicate when a package has reached the end of its intended life.
+
+The reason being is that the current package statuses mean specific things today and how they are perceived such as:
+
+- Vulnerable - Indicates the package has a known security vulnerability registered in an official CVE database.
+- Deprecated - Indicates the package is deprecated, meaning that people should consider migrating off of the package for a common reason such as it is **legacy, included critical bugs, or any other reason** the maintainer wants to disclose. These packages are no longer recommended to use but may still receive updates.
+
+Introducing a concept of **"end of life"** will make it clear to package authors that these packages are not intended to be used any further as they are no longer supported, will not receive critical updates, and feedback will not be considered. This would be a new definition including:
+
+- End of life (EOL) - Indicates the package is no longer maintained, supported, or updated by its authors.
+
+| Status      | Definition                              | Updates Available | User Guidance                          | Risk     |
+|-------------|-----------------------------------------|-------------------|----------------------------------------|----------|
+| EOL         | No longer maintained or supported       | No                | Migrate to alternatives                | High     |
+| Vulnerable  | Contains known security vulnerabilities | Possibly          | Update or find secure alternatives     | High     |
+| Deprecated  | Not recommended for use                 | Possibly          | Transition to newer packages           | Moderate |
```

Comment:
According to https://learn.microsoft.com/en-us/nuget/nuget-org/deprecate-packages

The "long description" of the legacy deprecation message is "this package is legacy and is no longer maintained".

Therefore, it appears to me that you're adding a new way to do the same thing.

I'm also concerned that even if you intend that deprecate and EOL have different definitions as per the spec, that few people read the docs and in everyday language, deprecated and end of life effectively mean the same thing making it a bit confusing.

Reply:
We absolutely do not want to re-invent the wheel here. But I do want to share how people perceive this and our current data (I'll add it into the spec as well):

Here is a quick pulse survey over the last day on how people perceive it as is:

<img width="593" alt="image" src="https://github.com/NuGet/Home/assets/1288848/1f3485b3-603e-4f49-a6cd-9bd862275539">

As for data, deprecated w/ legacy is the leading "reason".

<img width="807" alt="image" src="https://github.com/NuGet/Home/assets/1288848/d198a7e7-4892-4754-840a-32b738f5998e">

I believe we may need to find a clear separation and better flow for these package statuses so they do not come as a surprise to developers. A natural flow might be marking a package deprecated for a specific reason and after a specific timeframe, to mark it end of life to encourage people further to migrate.

I think Levi's comment above touches upon providing users a deprecation window before a package is then EOL'd.

Reply:
I can't help coming back to Andy's point: how is Deprecated different from EOL?  Would the workflow on https://learn.microsoft.com/en-us/nuget/nuget-org/deprecate-packages be any different for EOL (other than not needing a reason)?  And how does this proposal address the motivations differently from the Deprecated status, other than by exposing the state more forcefully in the client?

My main concern is that many of these proposed changes create an implied distinction that EOL is a terminal state, but Deprecated is not (contrary to how it is defined today).

Summarizing and addressing the Motivations:
* **Am I supported?** - this is already encompassed in the definition of being Deprecated.  While Deprecated doesn't only half says "there will be no support or updates", it's pretty clearly a declaration of "use this package at your own risk".
  > You can deprecate a package if you no longer maintain a package or if would like to encourage your package's consumers to move to another package.

  Deprecated is already indicated in some clients (such as VS, as shown in the documentation).  It sounds like this proposal would change from a warning triangle to a clearer "stop" signal, which seems to further artificially imply that Deprecated is not EOL.
  ![](https://learn.microsoft.com/en-us/nuget/nuget-org/media/deprecation-vs.png)
* **Security** - if consumers are not updating Deprecated packages, what will incentivize them by relabeling to EOL?  Just by treating it differently on the client?  Why not treat Deprecated packages that way on the client?  Will "EOL" be regarded differently from "Deprecated" (e.g. in the poll results)?
* **Pre-communicated lifetime** (added to incorporate Levi's excellent point) - Can we schedule a package to become Deprecated?  Does that reduce the value of a separate EOL state?

To me, the biggest potential value I see proposed here is
> A standalone feature also allows for more tailored experiences in the future without being constrained to the existing deprecation framework. As needs evolve for end of life, features can be easily adapted and extended.

However, this seems like "We want to change the deprecation experience, without changing it in situ", but in doing so creates a lot of ambiguity.  Is there a concrete scenario that isn't already somehow addressed by Deprecated?

Reply:
The core question here is whether the ecosystem has evolved to the point where it needs clearer definitions for these two states:

- Limited future support, where the package is still usable but not recommended for long-term use (Deprecated).

- No future support, where the package should no longer be used (End of Life).

I do not see this as a technology problem (i.e. does the technology/capabilities exist, yes, they do), but rather a matter of what definitions people would prefer to see and how they would be emphasized.

Reply:
Is there a demand for the limited future support state?  (Not doubting, just asking if there's data.)  In my limited experience, I've seen planned EOLs ("not supported after <date>"), but never encountered a package with the 3-phase lifecycle described below in the spec (`Package created -> package deprecated -> package end of life`) where a package has been marked deprecated but not EOL.

Reply:
> The challenges here is that deprecation today does not encourage the urgency warranted (their words that I also share) and we need to bring more awareness to when packages are deprecated / end of life.

This statement defines a slightly different problem: when deprecation is signaled, it isn't being treated as an urgent priority.  Would that be addressed by escalating how deprecation is represented to end users?  E.g. if we change from "warning" iconography to "error" iconography (as you've described for EOL), would it make deprecated packages seem more urgent?

Reply:
I believe the extra state is a good idea, since we're currently overloading the deprecated state to mean two things. Currently, when we want to move an Azure SDK to an EOL state, we have to maintain it for 3 years. During that period, the library is 'deprecated' but still supported. Once those three years are up, we'd like to EOL it, but we're relying on custom messages added to the 'deprecated' state to convey this. The process would be more crisply defined with a separate EOL state we could apply to packages that are outside that 3 year window.